src/operator/contrib/dgl_graph.cc

```diff
@@ -35,6 +35,833 @@
+  std::vector<dgl_id_t> tmp_edge_list;
+  std::vector<dgl_id_t> tmp_sampled_src_list;
+  std::vector<dgl_id_t> tmp_sampled_edge_list;
+  std::unordered_map<dgl_id_t, neigh_list> neigh_mp;
+  size_t num_edges = 0;
+  while (!node_queue.empty()) {
+    ver_node& cur_node = node_queue.front();
+    if (cur_node.level < num_hops) {
+      dgl_id_t dst_id = cur_node.vertex_id;
+      tmp_src_list.clear();
+      tmp_edge_list.clear();
+      tmp_sampled_src_list.clear();
+      tmp_sampled_edge_list.clear();
+      GetSrcList(val_list,
+                 col_list,
+                 indptr,
+                 dst_id,
+                 &tmp_src_list,
+                 &tmp_edge_list);
+      if (probability == nullptr) {  // uniform-sample
+        // Here we can also use GetUniformSampleShuffle() API
+        GetUniformSample(tmp_src_list,
+                       tmp_edge_list,
+                       num_neighbor,
+                       &tmp_sampled_src_list,
+                       &tmp_sampled_edge_list);
+      } else {  // non-uniform-sample
+        GetNonUniformSample(probability,
+                       tmp_src_list,
+                       tmp_edge_list,
+                       num_neighbor,
+                       &tmp_sampled_src_list,
+                       &tmp_sampled_edge_list);
+      }
+      neigh_mp.insert(std::pair<dgl_id_t, neigh_list>(dst_id,
+        neigh_list(tmp_sampled_src_list, tmp_sampled_edge_list)));
+      num_edges += tmp_sampled_src_list.size();
+      bool exit = false;
+      for (size_t i = 0; i < tmp_sampled_src_list.size(); ++i) {
+        auto ret = sub_ver_mp.insert(tmp_sampled_src_list[i]);
+        if (ret.second) {
+          sub_vertices_count++;
+          ver_node new_node;
+          new_node.vertex_id = tmp_sampled_src_list[i];
+          new_node.level = cur_node.level + 1;
+          if (layer.find(tmp_sampled_src_list[i]) == layer.end()) {
+            layer[tmp_sampled_src_list[i]] = new_node.level;
+          }
+          if (new_node.level < num_hops) {
+            node_queue.push(new_node);
+          }
+          if (sub_vertices_count == max_num_vertices) {
+            exit = true;
+            break;
+          }
+        }
+      }
+      if (exit) {
+        break;
+      }
+    }
+    node_queue.pop();
+  }
+  // Copy sub_ver_mp to output[0]
+  size_t idx = 0;
+  for (auto& data : sub_ver_mp) {
+    *(out+idx) = data;
+    idx++;
+  }
+  size_t num_vertices = sub_ver_mp.size();
+  std::sort(out, out + num_vertices);
+  // The rest data will be set to -1
+  for (dgl_id_t i = idx; i < max_num_vertices; ++i) {
+    *(out+i) = -1;
+  }
+  // The last element stores the actual
+  // number of vertices in the subgraph.
+  out[max_num_vertices] = sub_ver_mp.size();
+  // Copy sub_probability
+  if (sub_prob != nullptr) {
+    for (dgl_id_t i = 0; i < max_num_vertices; ++i) {
+      dgl_id_t idx = out[i];
+      if (idx != -1) {
+        sub_prob[i] = probability[idx];
```

Comment:
we need to have the probability that a neighbor is sampled from the the neighbor list. In other words, it should be normalized for the neighbor list.